_includes/nav.html

```diff
@@ -1,73 +1,99 @@
-  depending on their `nav_order` value.
+  First, the array of `nav_order_pages` is divided into `number_order_pages` and
+  `string_order_pages` according to the type of `nav_order` value.
   
-  The first character of the `jsonify` result is `"` only for strings.
+  The first character of the result of `jsonify` is `"` only for strings.
+  Grouping by a single character also ensures the number of groups is small.
 {%- endcomment -%}
 
-{%- assign nav_ordered_groups = nav_ordered_pages
+{%- assign number_order_pages = "" | split: "" -%}
+{%- assign string_order_pages = "" | split: "" -%}
+{%- assign nav_order_groups = nav_order_pages
       | group_by_exp: "item", "item.nav_order | jsonify | slice: 0" -%}
-
-{%- assign number_ordered_pages = "" | split: "" -%}
-{%- for group in nav_ordered_groups -%}
+{%- for group in nav_order_groups -%}
   {%- if group.name == '"' -%}
-    {%- assign string_ordered_pages = string_ordered_pages | concat: group.items -%}
+    {%- assign string_order_pages = string_order_pages | concat: group.items -%}
   {%- else -%}
-    {%- assign number_ordered_pages = number_ordered_pages | concat: group.items -%}
+    {%- assign number_order_pages = number_order_pages | concat: group.items -%}
   {%- endif -%}
 {%- endfor -%}
 
-{%- assign sorted_number_ordered_groups = number_ordered_pages
-      | sort: "nav_order" | group_by: "nav_order" -%}
+{%- assign sorted_number_pages = number_order_pages | sort: "nav_order" -%}
 
 {%- comment -%}
-  Group the string-ordered pages by `nav_order`, if non-nil, and otherwise `title`
-  (but appending the empty string to a numeric title to convert it to a string).
-  
-  Then sort the groups according to the site setting for case (in)sensitivity.
+  The `title_order_pages` and the `string_order_pages` need to be sorted together.
+  When either or both of them is empty, `sorted_string_pages` is simple to compute.
+  Otherwise, we need to group their concatenation by their `nav_order` or
+  `title` values, sort the groups by those values, and concatenate the grouped
+  arrays of pages.
 {%- endcomment -%}
 
-{%- assign string_ordered_groups = string_ordered_pages
-      | group_by_exp:"item", "item.nav_order | default: item.title | append: '' " -%}
+{%- if title_order_pages == empty and string_order_pages == empty -%}
+
+  {%- assign sorted_string_pages = string_order_pages -%}
+
+{%- elsif title_order_pages == empty -%}
+
+  {%- if site.nav_sort == 'case_insensitive' -%}
+    {%- assign sorted_string_pages = string_order_pages | sort_natural: "nav_order" -%}
+  {%- else -%}
+    {%- assign sorted_string_pages = string_order_pages | sort: "nav_order" -%}
+  {%- endif -%}
+
+{%- elsif string_order_pages == empty -%}
+
+  {%- if site.nav_sort == 'case_insensitive' -%}
+    {%- assign sorted_string_pages = title_order_pages | sort_natural: "title" -%}
+  {%- else -%}
+    {%- assign sorted_string_pages = title_order_pages | sort: "title" -%}
```

Comment:
FYI - this is the line triggering the actions failure!

Reply:
> there's some mismatch between the deployment used for regression tests and ours (ex: different Jekyll versions), and that's causing us to miss this error

Indeed! When I checked my terminal log, I discovered that I was using Jekyll 4.2.2. I usually change the `jekyll` constraint in `just-the-docs.gemspec` from `">= 3.8.5"` to "~> 3.8.5"`, to avoid that happening. After I did that and updated the bundle, my local build fails in the same way.

I suggest to remove `Gemfile.lock` from `.gitignore`, so that `bundle install` will always use exactly the same gem versions when building the theme website. `Gemfile.lock` is _not_ in [the official Jekyll `.gitignore`](https://github.com/github/gitignore/blob/main/Jekyll.gitignore) – presumably for good reasons…

For the regression tests, I do check in the `Gemfile.lock`, and I get `Using jekyll 3.8.7`. (I think the `github-pages` gem requires jekyll 3.9.0, but I'm not using that.) 

I'm not yet sure of the cause of the Jekyll 3 build failure, but I seem to recall that Jekyll 4 made some subtle breaking changes around the treatment of `nil` by filters. I suspect the version of Liquid used by Jekyll 4 also differs regarding coercions between numbers and strings, although I've never managed to pinpoint how. I'm surprised that the relatively rich navigation hierarchy on the tests site didn't reveal the bug.

Reply:
@mattxwang Presumably our `.gitignore` and `Gemfile.lock` files don't affect users of the theme, so we could defer discussion of whether to do that to after the release of v0.4.0. But as this comment is hidden in a merged PR, perhaps I should submit a fresh PR for the relevant updates, to ensure it doesn't get overlooked?

> > there's some mismatch between the deployment used for regression tests and ours (ex: different Jekyll versions), and that's causing us to miss this error
> 
> Indeed! When I checked my terminal log, I discovered that I was using Jekyll 4.2.2. I usually change the `jekyll` constraint in `just-the-docs.gemspec` from `">= 3.8.5"` to "~> 3.8.5"`, to avoid that happening. After I did that and updated the bundle, my local build fails in the same way.
> 
> I suggest to remove `Gemfile.lock` from `.gitignore`, so that `bundle install` will always use exactly the same gem versions when building the theme website. `Gemfile.lock` is _not_ in [the official Jekyll `.gitignore`](https://github.com/github/gitignore/blob/main/Jekyll.gitignore) – presumably for good reasons…